content/blog/biscuit-3-3.md

```diff
@@ -0,0 +1,182 @@
+
+Libraries are supposed to encode tokens with the smallest version number possible, in order to facilitate gradual migration.
+
+### Signed block version number
+
+Starting with biscuit `v3.3`, the spec also defines a version number for the block signatures. This will allow improving signatures in a more graceful way.
+
+## tl;dr: partial breaking changes
+
+Biscuit has a strong policy for making spec updates additive:
+
+- tokens emitted with a library supporting biscuit 3.0 to 3.3 will be handled correctly by a library supporting biscuit 3.3;
+- tokens emitted by a library supporting biscuit 3.3, but not using any features from biscuit 3.3 will be handled correctly by a library supporting the features they use.
+
+However, biscuit 3.2 introduced a breaking change for third-party blocks, making third-party blocks emitted with a 3.0/3.1 library not accepted anymore.
+
+Biscuit 3.3 also introduces a breaking change on third-party blocks. New third-party blocks will not be supported by biscuit 3.2 libs, and third-party blocks emitted with biscuit 3.2 will be rejected by default by 3.3 libs (this can be relaxed during a migration period).
+
+### Datalog syntax changes
+
+The spec guarantees that datalog updates are purely additive, when encoded in tokens.
+
+The textual syntax for datalog has been updated in biscuit 3.3:
+
+- sets are now delimited by `{}`
+- strict equality tests are now denoted by `===` and `!==`
+
+## Datalog improvements
+
+### Arrays and maps / JSON support
+
+Up until now biscuit datalog only had a single collection type: sets. Sets were quite restrictive (no nesting, homogeneous). This made impossible for a biscuit token to carry JSON for instance.
+
+Biscuit 3.3 adds support for arrays, maps, and `null`, thus providing a way to embed arbitrary JSON values in a token.
+
+```
+payload({"key": ["value", true, null, {}]});
+```
+
+#### `null`
+
+Arrays and map support `.get()`, so we needed a way to handle missing keys. Since biscuit datalog is untyped, `null` is an okay solution for this. `null` was also the last missing piece for JSON support.
+
+```
+check if [].get(0) == null;
+```
+
+#### Closures
+
+With this new focus on collection types, we needed a way to express more things in the language. Datalog expressions follow a pure evaluation model, so mutability and loops were not available. Higher-order functions were thus the best way to work with collections.
+
+Arrays, sets and maps support `.any()` and `.all()`, taking a predicate. Closures are not first-class (meaning they cannot be manipulated like regular values), but can however be nested (to work with nested data types). Variable shadowing was not possible until then (since all variables could only be bound in the same scope, with predicates). Variable shadowing is now possible syntacticly, but explictly forbidden by the spec and rejected by implementations.
```

Comment:
```suggestion
Arrays, sets and maps support `.any()` and `.all()`, taking a predicate. Closures are not first-class (meaning they cannot be manipulated like regular values), but can however be nested (to work with nested data types). Variable shadowing was not possible until then (since all variables could only be bound in the same scope, with predicates). Variable shadowing is now possible syntactically, but explicitely forbidden by the spec and rejected by implementations.
```

Reply:
(explicitly, not explicitely btw)